src/algebra/monoid_algebra.lean

```diff
@@ -373,6 +377,81 @@ end
 
 end misc_theorems
 
+/-! #### Non-unital, non-associative algebra structure -/
+section non_unital_non_assoc_algebra
+
+variables (k) [semiring k] [has_mul G]
+
+instance is_scalar_tower_self : is_scalar_tower k (monoid_algebra k G) (monoid_algebra k G) :=
+⟨λ t a b,
+begin
+  ext m,
+  simp only [mul_apply, finsupp.mul_sum, smul_apply, sum_smul_index, implies_true_iff,
+    eq_self_iff_true, smul_eq_mul, sum_zero, zero_mul, mul_assoc, mul_ite, mul_zero, if_t_t],
+end⟩
+
+/-- If the coefficients are commutative amongst themselves, the also commute with the algebra
+multiplication. -/
+instance smul_comm_class_self {k : Type u₁} [comm_semiring k] :
+  smul_comm_class k (monoid_algebra k G) (monoid_algebra k G) :=
+⟨λ t a b,
+begin
+  ext m,
+  simp only [mul_apply, finsupp.mul_sum, smul_apply, sum_smul_index, implies_true_iff,
+    eq_self_iff_true, smul_eq_mul, mul_ite, mul_zero, mul_left_comm, ite_eq_right_iff],
+end⟩
+
+instance smul_comm_class_symm_self {k : Type u₁} [comm_semiring k] :
+  smul_comm_class (monoid_algebra k G) k (monoid_algebra k G) :=
+smul_comm_class.symm k (monoid_algebra k G) (monoid_algebra k G)
+
+/-- The functor `G ↦ monoid_algebra k G`, from the category of magmas to the category of non-unital,
+non-associative algebras over `k` is adjoint to the forgetful functor in the other direction. -/
+@[simps] def lift_magma {A : Type u₃}
```

Comment:
Unfortunately the fact that `A` is non-unital makes it hard to factor my work through `lift_nc` (more precisely, a generalisation of `lift_nc` that relaxes its `[semiring R]` argument to `[non_unital_non_assoc_semiring R]`).

The reason is that the defining property of `lift_nc` is:
```lean
@[simp] lemma lift_nc_single (f : k →+ R) (g : G →* R) (a : G) (b : k) :
  lift_nc f g (single a b) = f b * g a :=
...
```
whereas the defining property of `lift_magma` is:
```lean
variables {A : Type u₃}
variables [non_unital_non_assoc_semiring A] [module k A] [is_scalar_tower k A A] [smul_comm_class k A A]
example (g : mul_hom G A) (a : G) (b : k) :
  lift_magma k g (single a b) = b • g a :=
...
```
I claim that to unify these we need a unit in `A` since we'd need to invoke `lift_nc` using `λ (b : k), (b • (1 : A))` for `f`.

I agree that with sufficient effort some unification between `lift_nc` and `lift_magma` could be obtained but the above makes me doubt that it is worth it.

Reply:
That's a good point - I think we can still relax to remove associativity requirements from lift_nc, but I guess that's not in scope here.

Reply:
Hmm, actually `lift_nc` seems to do exactly what you need here if you weaken its arguments to:
```lean
def lift_nc {R : Type*} [non_unital_non_assoc_semiring R] (f : k →+ R) (g : mul_hom G R) : monoid_algebra k G →+ R :=
lift_add_hom (λ x : G, (add_monoid_hom.mul_right (g x)).comp f)
```
You can then probably weakenn the subsequent lemmas to match.

Reply:
I'm sorry if I'm being dense here but I still don't see how to use `lift_nc` in `lift_magma`.

The problem is that when we are in the setting of `lift_magma`, because `A` has no unit, there is no way to construct the map `f` demanded by `lift_nc`.

I also note (though only as a secondary issue) that even without this problem, we could not use `lift_nc_mul` because it requires the multiplication in `R` to be associative (or at least to satisfy `mul_mul_mul_comm`).

I haven't looked through all the uses of `lift_nc` but I wouldn't be surprised if the right way resolve the overlap of responsibilities between `lift_magma` and `lift_nc` was to drop `lift_nc` entirely and make `lift_magma` the fundamental definition. In other words, the fundamental object into which we lift is a `non_unital_non_assoc_semiring R` but where `R` carries a compatible scalar action (as in `lift_magma`) rather than carrying a map `f : K → R` (as in `lift_nc`).

Please do enlighten me if I'm missing the obvious here.

Reply:
I suspect it's me being dense after all - how about I make a PR with the generalizations I'm suggesting tomorrow, and then I'll think again about whether they're actually relevant here?

Reply:
By all means, if you have the appetite. I'll push on with some other work in an independent direction in the meantime. Thanks for all the great suggestions.